### ISSUE TITLE
Accessing Query.selectable causes RecursionError in "before_compile" event listener

### ISSUE THREAD

**Issue by the-wondersmith** [sqlalchemy/sqlalchemy#5487]
I posted a [question](https://stackoverflow.com/questions/63180197/sqlalchemy-change-table-name-in-fromclause-w-event-listener) on StackOverflow pertaining to what I'm actually trying to accomplish with the `before_compile` event listener. In the course of experimenting while waiting for an answer, I made some headway but eventually got stuck.

It seems as though attempting to access virtually any attribute or property of the Query object passed into the registered event handler causes the interpreter to throw a recursion error. I can get at the Query object's attributes if I attach a debugger and set a break point on the function entry, so I know that it at least *appears* to have all of the attributes that I'd like to modify prior to compilation, I just can't... do anything with them.

I've tried to dig into this and the best I can tell is that accessing certain attributes causes the event handler to be fired again? Either that or there are multiple circular references between the various things that comprise the Query objects attributes and properties.

The full explanation of why I'd like to do so is in the StackOverflow link above, but the much condensed version is that I'm stuck supporting a very silly DB design that breaks out what should be one table into ~28 different tables that all share the same layout. Due to other entangled complications, SQLAlchemy's table inheritance recipes don't feel like the best option. The best thing I can think of would be to utilize an event listener to modify emitted queries prior to their compilation and set the correct table name for the record that's being requested.

What did I miss in the documentation about how the `before_compile` event works?

**Comment by zzzeek** [comment]
> I posted a [question](https://stackoverflow.com/questions/63180197/sqlalchemy-change-table-name-in-fromclause-w-event-listener) on StackOverflow pertaining to what I'm actually trying to accomplish with the `before_compile` event listener. In the course of experimenting while waiting for an answer, I made some headway but eventually got stuck.

So what that example is doing is known as horizontal sharding.   SQLAlchemy has a horizontal sharding extension that is oriented around having different database connections for different criteria.    Since it seems you are using an ODBC driver that can access per-individual file I would most likely look to just using this extension:

https://docs.sqlalchemy.org/en/13/orm/examples.html#examples-sharding

then you would use 28 different engines you can create dynamically.

I'm not sure how this leads to wanting to use a different "tablename" for each, if that's some requirement of how this has to work then I would suggest, as the above document also does, to use the entity name pattern: https://github.com/sqlalchemy/sqlalchemy/wiki/EntityName   

which incidentally can be used for the multiple database approach also.    you might also be interested in being able to customize which "bind" the Session uses on a per-query basis, which is how the horizontal shard extension works, an example of that is at https://docs.sqlalchemy.org/en/13/orm/persistence_techniques.html#partitioning-strategies-e-g-multiple-database-backends-per-session .


> 
> It seems as though attempting to access virtually any attribute or property of the Query object passed into the registered event handler causes the interpreter to throw a recursion error. I can get at the Query object's attributes if I attach a debugger and set a break point on the function entry, so I know that it at least _appears_ to have all of the attributes that I'd like to modify prior to compilation, I just can't... do anything with them.

The term "virtually any attribute or property" appears to be hyperbole to me, since of course the Query object can be accessed within the before_compile() event, it's [right in the examples](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=before_compile#sqlalchemy.orm.events.QueryEvents.before_compile), and it would be more useful if you provided specifics.      However, If I had to guess, it is likely that you are making use of a function like ``query.subquery()`` or ``query.from_self()`` that is itself going to compile itself in the 1.3 series, and if your own before_compile() handler is not ready to anticipate this, then you will in fact have a recursion error.   

There is a way to write the event handler such that it anticipates such a reentrant condition, however, I think you should stick with the prior mentioned sharding approaches using either horizontal sharding or simple entity name.

Also, the before_compile() event handler won't do this in 1.4 and in fact the whole event will be gone by SQLAlchemy 2.0 and replaced with much more effective systems that will be available in the 1.4 release.

**Comment by the-wondersmith** [comment]
@zzzeek  You're correct, my usage of the phrase "virtually any attribute or property" was indeed hyperbole. I was frustrated, and I apologize.

I did look at the horizontal sharding / EntityName patterns before I did anything else, and while you're absolutely correct that they would be adaptable to my use case, they don't feel like the right solution for my (admittedly esoteric) situation.

The legacy application I'm wrapping / extending is (essentially) a warehouse inventory management software. The problematic tables contain all of the customer-specific records. I've already implemented a scoped session wherein each relevant model is bound to the correct session. It looks something like [this](https://gist.github.com/the-wondersmith/58ca1cedfc814a233961a778ddd88da2#file-example-py).

I've also tested that what I'd *like* to do is both possible and works by using a `before_cursor_execute` listener bound to the engine dedicated to the customer records. It works by having the `__tablename__` attribute on the Customer model set to `ƒ` (a character which absolutely will not ever show up in a real query otherwise) and then just using some convoluted logic and the standard `str.replace()` method. It does in fact work, but the implementation is *ugly as sin* and probably more brittle than balsa wood. It looks something like [this](https://gist.github.com/the-wondersmith/58ca1cedfc814a233961a778ddd88da2#file-table_fix-py).

I'm fairly certain that setting the `__tablename__` in the `Query` object prior to compilation would be a much more stable, reliable, elegant way to do what I'm trying to do. When I try to do so though (and apologies again for my earlier hyperbole) I trigger a `RecursionError` any time I try to access `Query.selectable` or `Query.statement`. FWIW, I don't seem to have issues accessing `Query.whereclause` though.

The models in question do have foreign keys and relationships, but are specifically configured to lazy load and to emit select statements for them so that I can be reasonably sure that I *won't* generate any joins or subqueries as the underlying ODBC driver isn't the best at handling them. That is to say, I'm not trying to use `query.subquery()` or `query.from_self()` as I didn't think either one would allow me to modify the table name being compiled into the to-be-emitted statement. I figured that what I wanted would be somewhere in either the `selectable` or `statement` properties.

What did I miss?

**Comment by zzzeek** [comment]
The `__tablename__` attribute is only used when classes are mapped, and after that, it is no longer consulted.  the ".name" field of `Table`, which is where it goes, cannot be changed.   This name is linked to how the `Table` object is stored and accessed within its `MetaData` collection and is also linked to SQL statements that are going to be cached.  `Table` object is a module-level object that is shared between application threads, so it's not appropriate for mutable state in any way.

any changes to the SQL that's rendered therefore have to be local to that query.  So doing this at the ORM/object level would mean that you'd make use of the entity name pattern, which does not lend itself to occurring in an "implicit" way; you'd need to use the correct ORM entity up front, something like:

   entity = Customer.for_name("alfred")
   session.query(entity).filter(entity.foo == 'bar').first()

> When I try to do so though (and apologies again for my earlier hyperbole) I trigger a RecursionError any time I try to access Query.selectable or Query.statement. FWIW, I don't seem to have issues accessing Query.whereclause though.

This is because .selectable and .statement are intended to be used when one wishes to `.compile()` the query explicitly.   There is no valid use case for accessing these attributes inside of the before_compile() event, because in this event, you're already inside the process of `.selectable` or `.statement` being invoked already.    Accessing these attributes does not provide any path to altering the state of the Query itself within the event hook.  The only way to alter the Query in the event hook in this way would be to fully rewrite the Query entirely using a new entity via the entity name approach above.   Query is not designed for this level of mutation, and in fact that's part of why it's going away in SQLAlchemy 2.0 in any case to be fully replaced by the select() construct that does allow this degree of mutation using a replacement traversal.     none of this is feasible within the 1.3 series however.

the before_cursor_execute() approach is what I'd recommend in this case, in fact, short of the availability of new features in 1.4 that will allow this to be done at the statement manipulation level more effectively.   SQLAlchemy has a feature that works like this already called "schema translate", which applies to the "schema" portion of the table.   

I haven't looked at your approach, however, you should try to follow along the example in the Wiki that is in fact doing exactly what you are referring towards here, that is, changing the names of tables on a per-Session basis; see the example at https://github.com/sqlalchemy/sqlalchemy/wiki/SessionModifiedSQL

**Comment by the-wondersmith** [comment]
@zzzeek That gets me closer to what I need, for sure. 

The issue I'm running into (mentally at least) is that a given session *might* only be accessing a single customer record, in which case doing the table modification the same way that Wiki example does is totally feasible. It's not possible to know in advance if a given session is going to need to access more than one of the customer tables though.

There are some strange internal rules in the legacy application that govern how the customer records are sharded, but for simplicity's sake let's say that:

- All customer records are sorted into tables based on the first letter of their last name, with table names formatted like `NAMEP{letter}`
- All customer tables have an `AcctNumber` column that wherein the first two digits of a given customer's `AcctNumber` corresponds to the ASCII character code for the first letter of their last name
  - I.E. `Johnny Appleseed`s `AcctNumber` will begin would be something like `65123456`
- The `AcctNumber` column also exists in any other table that needs to link a given customer to a given record on that table

Therefore, it's possible to figure out which table needs to be queried by examining a query to see if it contains a usable value for either `Customer.LastName` or `Customer.AcctNumber`.

Would it be possible to introspect a query object to check for the presence of referenceable values for either field and then pass that information down to the `before_cursor_execute` listener using the `session.info` property?

I suppose a better way to phrase my conundrum is that although the best (if not only) place to set the table name would be in the `before_cursor_execute` hook, the best place to *figure out* which table needs to be queried is *before* it's actually compiled so that I don't have to effectively undo that compilation in order to do the figuring. Does that make sense?

Also, side question, if SQLAlchemy 1.4 has better tools for this, is there a publically available version I could experiment with?

**Comment by zzzeek** [comment]
yes I really think you need to be using the horizontal sharding extension for this.   in order to check the fields of Query you look at [.column_descriptions](https://docs.sqlalchemy.org/en/13/orm/query.html?highlight=column_descriptions#sqlalchemy.orm.query.Query.column_descriptions) and [.whereclause](https://docs.sqlalchemy.org/en/13/orm/query.html?highlight=column_descriptions#sqlalchemy.orm.query.Query.whereclause).  An example of this is in the hoirizontal shard demo at https://docs.sqlalchemy.org/en/13/_modules/examples/sharding/attribute_shard.html,where it's using query._criterion, same thing as whereclause, that's a very in-depth example.

in order to adapt horizontal sharding to single-engine with different table names, make copies of the engine with different execution options:

shards = {
   "shard1": engine.execution_options{shard="shard1"),
   "shard2": engine.execution_options{shard="shard2"),
   "shard3": engine.execution_options{shard="shard3"),
   ...
}

then use the before_cursor_execute event as before, and exract the information that leads you to the desired table name by looking at the "context" argument passed for the execution options, it should be in a few places including context.execution_options .     here's an example of this:  https://github.com/sqlalchemy/sqlalchemy/blob/master/test/ext/test_horizontal_shard.py#L719

**Comment by the-wondersmith** [comment]
@zzzeek I followed along with the examples you gave and I've got something reliable worked out. I'll post a gist for it shortly and update this issue with the link when I do, in case anyone else runs into the same issue in the future.

I've run into a fun side issue though that stems from the underlying ODBC driver. The SQLALchemy Dialect I'm using is one that I've personally written, though @gordthompson has given me a considerable amount of direction (some personally and some by virtue of being the author of sqlalchemy-access). The dialect appears to handle absolutely everything correctly,but the Microsoft Paradox ODBC driver apparently throws a fit whenever a column name containing the `#` character is passed in for an `INSERT`.

There is an extremely simple work around for this though:

1) Omit the column(s) and values for any column whose name contains `#` from the insert statement
2) Execute the modified insert statement as normal
3) Immediately follow the execution of the modified insert statement with an update statement for the omitted columns and values, using the non-omitted values in the `WHERE` clause

**EX:**

> Unmodified Insert Statement
> ```sql
> INSERT INTO "CUSTOMERS_A" ("Acct #", "First Name", "Last Name") VALUES (123456, 'Johnny', 'Appleseed');
> ```

> "Corrected" Workaround Routine
> ```sql
> INSERT INTO "CUSTOMERS_A" ("First Name", "Last Name") VALUES ('Johnny', 'Appleseed');
> UPDATE "CUSTOMERS_A" SET "Acct #" = 123456 WHERE "First Name" = 'Johnny' AND "Last Name" = 'Appleseed';
> ```

I think the best place to handle this in the execution flow would be in the dialect's `DefaultExecutionContext` by overriding the `pre_exec` and `post_exec` methods wouldn't it? Or is there a better option somewhere else?

**Comment by gordthompson** [comment]
One potential problem with that workaround would be if "Acct #" was the primary key. Does Paradox support INSERTs into views? If so, then you could try creating a temporary view ...

``` sql
CREATE VIEW zzz_temp AS SELECT "Acct #" AS replacement1, "First Name", ... FROM "CUSTOMERS_A"
```

and then do

``` sql
INSERT INTO zzz_temp (replacement1, "First Name", ...) VALUES ...
```

**Comment by zzzeek** [comment]
if the database cannot support INSERT into a column with a pound sign in it, then that database simply does not support column names with pound signs.  If not done so already, I would look to see if there's a quoting syntax that makes it happy, like using brackets, or something.  how does this legacy database actually exist with this column name and it has data in it?   the  creators of this DB never used an INSERT statement ?

**Comment by gordthompson** [comment]
>  the creators of this DB never used an INSERT statement ?

That is quite possible. Using the Access UI (MSACCESS.EXE) I can create an Access database, create a table in that database, and add a row to that table without any CREATE DATABASE, CREATE TABLE, or INSERT statements being used. The Access UI takes my pointy-clicky input and uses Microsoft Jet/ACE DAO (Data Access Objects) to perform those tasks.

Access SQL is just a layer on top of DAO. Similarly Paradox ODBC could just be a layer on top of its underlying transactional engine (record manager). Paradox and Access were "born" about the same time and were probably somewhat similar in their architecture.

**Comment by zzzeek** [comment]
so these paradox DB files are like, passed around your company already and folks would not be able to have this data in some other format?   this is such a lot of effort for something that could just be stuck into a SQLite file and be done with it.

**Comment by zzzeek** [comment]
it's also not unheard of to have SQLite act as a "cache" for data that's in some other format; pandas has some stuff like this going on I believe.

**Comment by the-wondersmith** [comment]
@zzzeek  I have looked high and low for a quoting syntax that will work, but it just doesn't exist. The Paradox driver is included in the Microsoft ODBC system by default for reasons that are absolutely beyond me, but it is a convenient solution to my problem. I honestly have *zero* idea what the original DB creators did or were thinking, but you're 100% correct - the driver will not allow column names with `#` to be present in an insert statement, period. It doesn't seem to care one whit about them being in select, update, or delete statements though.

The legacy app that I'm wrapping / extending was written to use the [Borland Database Engine](https://en.wikipedia.org/wiki/Borland_Database_Engine). I suspect that the root of the issue is that it pre-dates what you can actually find about the BDE nowadays, and while the BDE is *compatible* the app is *actually* written to just use Paradox tables, not the actual BDE.

I don't actually have anything whatsoever to do with the legacy app itself, but I am in contact with the original creator and he is absolutely willing to tell me anything I need to know about the internal workings of the app itself and how things are stored / created. A more modern version of the app just isn't in the cards though, and it absolutely cannot support what I need to do with it.  I don't disagree with you *at all* that a SQLite instance would be a way easier solution, but in this instance there's no way around it - I need two-way data transfer between the app's backend and my wrapper.

I had thought about using something like [pypxlib](https://github.com/mherrmann/pypxlib) to supplement the gaps / sticky bits in the ODBC driver, but truth be told there aren't *actually* any gaps in the ODBC driver's functionality, nor are there any *real* gaps in my SQLAlchemy dialect's support thereof. It's just this one edge case that honestly *shouldn't* exists, but does anyway.

@gordthompson is completely correct about how the Paradox ODBC driver works, it's just another layer on top of the JET engine. In fact, I've actually had a similar issue with an Access DB throwing the same issue. I was able to work around that by simply omitting the offending column from ever appearing any of the queries I ended up generating, as it didn't hold anything relevant to what I was doing. That's just not possible here.

I suppose my question is:

1) If I wanted to remove one or more pairs of columns and values from a query prior to its execution, at what point in the control-flow of a SQLAlchemy dialect is the best place to do so?
2) Given that doing this is absurd to begin this, there can't possibly be said to be a "proper" way to do so, but even so - what's your recommendation on the *best* way to do so?

**Comment by gordthompson** [comment]
Just curious: If you create an Access database and add an ODBC linked table that points to the Paradox table can you insert directly to the [Acct #] column of the linked table using Access ODBC?

**Comment by the-wondersmith** [comment]
@gordthompson That's a great idea but *oof*, I have no idea where to even start trying to check. I know off hand that I don't have a copy of Access lying around or installed, and neither do any of the systems this wrapper is intended to run on. 

Is that doable / checkable just using the ODBC subsystem? Or would I need a full-blown Access instance to check?

**Comment by gordthompson** [comment]
It can be done with just the Access Database Engine. Can you create an ODBC System DSN named aaaPdoxTest that points to the Paradox database folder? (Also, just to confirm that this is 32-bit, yes?)

**Comment by the-wondersmith** [comment]
It is 32-bit yes, and the way that I've set everything up in my testing / dev environment uses a system DNS pointed to the Paradox table files.

Also in case it helps, [here is all the info](https://gist.github.com/the-wondersmith/fcd53afc1b377cdaa9c3e3e3e527031f) that PyODBC returns when you interrogate the driver about what it supports.

**Comment by gordthompson** [comment]
So after thinking about this a bit more I'm not terribly optimistic that it will work, bit it's so easy that it's at least worth a try.

The script to create the Access database with a linked table is ...

``` python
import win32com.client  # needs `pip install pywin32`

# <replace with your values>
odbc_dsn_name = "my_paradox_dsn"
table_name = "CUSTOMER_A"
access_db = r"C:\Users\Gord\Desktop\gh_5487.mdb"
# </replace with your values>

dbLangGeneral = ";LANGID=0x0409;CP=1252;COUNTRY=0"
dbVersion40 = 64  # Access 2000/2003

dbe = win32com.client.Dispatch("DAO.DBEngine.36")
db = dbe.CreateDatabase(access_db, dbLangGeneral, dbVersion40)

tbd = win32com.client.Dispatch("DAO.TableDef.36")
tbd.Name = table_name
tbd.Connect = f"ODBC;DSN={odbc_dsn_name}"
tbd.SourceTableName = table_name
db.TableDefs.Append(tbd)
```

... and then you can test it with a script like ...

``` python
import pyodbc

# <replace with your values>
table_name = "CUSTOMER_A"
access_db = r"C:\Users\Gord\Desktop\gh_5487.mdb"
# </replace with your values>

pyodbc.pooling = False  # required for ODBC linked tables

cnxn = pyodbc.connect(
    "DRIVER={Microsoft Access Driver (*.mdb)};"
    f"DBQ={access_db}"
)
crsr = cnxn.cursor()
sql = f"""\
INSERT INTO [{table_name}] ([Acct #], [First Name], [Last Name])
VALUES (?, ?, ?)
"""
params = ("Acct001", "Gord", "Thompson")
crsr.execute(sql, params)
cnxn.commit()
```

**Comment by the-wondersmith** [comment]
@gordthompson  It was a good thought, but no joy it looks like. Failed with:

```
pywintypes.com_error: (-2147352567, 'Exception occurred.', (0, None, "ODBC--connection to 'paradox_access_test' failed.", 'jeterr40.chm', 5003151, -2146825137), None)
```

**Comment by the-wondersmith** [comment]
@zzzeek and @gordthompson  I've been poking at the Dialect I wrote for the MS Paradox driver, trying to find the best place to handle the "query splitting" I need to implement.

As best I can see, it would either be somewhere in the dialect's override of `compiler.SQLCompiler` or `default.DefaultExecutionContext`, but I can't quite tell which. If I implement it as part of the `DefaultExecutionContext`, that gives me logical places to run the split logic as well as somewhere to run the generated update statement (namely the `pre_exec` and `post_exec` methods).

However, doing that creates the unique complication of needing to effectively decompile the already-compiled statement so that I can correctly strip out any columns and values that are going to cause an error. Also, (at least in my case) one of the error-causing columns is actually one that SQLAlchemy regards as being an auto-increment primary key even though the underlying Paradox table holds no such distinctions. That makes me think that I may want to rope in the `SQLCompiler` class to help out, or perhaps perform the actual "split" somewhere higher up the chain than `DefaultExecutionContext` so that I can still check if a column needs a value pre-fetched (in my case it's actually pre-generated).

Basically, what I need to happen is for SQLAlchemy to emit queries as normal but for the Paradox dialect to catch inserts and transform them into two queries. Obviously this behavior is abnormal in the extreme, so it has no business being in SQLAlchemy itself as it's very much dialect-specific, which means it needs to be implemented in the dialect itself. Seeing as I wrote the dialect to being with, I'm happy to do the work to make it function the way I need it to. I'm just completely lost as to *where* in the dialect is the best place to accomplish it.

Thoughts?

**Comment by the-wondersmith** [comment]
@gordthompson  and @zzzeek  I think I've got it figured out. Thank you both profusely for your help. 

If anyone else runs into a similar issue, I've posted a gist detailing the solution I worked out [here](https://gist.github.com/the-wondersmith/92be7ff0b5954d612da35ead30768c47).
